File: lintgen/generated_agent/qwen-full/qwen2.5:32b-instruct/base+docs_agent/AndroidAutoDetector:InvalidUsesTagAttribute/iter_8.java
package com.android.tools.lint.checks;

import com.android.SdkConstants;
import com.android.annotations.NonNull;
import com.android.tools.lint.detector.api.*;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

import java.util.List;

public class AndroidAutoDetector extends XmlScanner {
    private static final String AUTOMOTIVE_APP_TAG = "automotiveApp";
    private static final String USES_TAG = "uses";
    private static final String NAME_ATTRIBUTE = "name";

    public static final Issue ISSUE = Issue.create(
            "InvalidAndroidAutoUsesName",
            "The 'uses' element in <automotiveApp> should contain a valid value for the 'name' attribute. Valid values are 'media', 'notification', or 'sms'.",
            "The `<uses>` element within an `<automotiveApp>` tag must have a `name` attribute with one of the following values: `media`, `notification`, or `sms`. This ensures that your app is correctly configured for Android Auto.",
            Category.CORRECTNESS,
            5,
            Severity.ERROR,
            new Implementation(
                    AndroidAutoDetector.class,
                    EnumSet.of(Scope.RESOURCE_FILE_SCOPE)
            )
    );

    @NonNull
    @Override
    public List<String> getApplicableElements() {
        return List.of(AUTOMOTIVE_APP_TAG);
    }

    @Override
    public void visitElement(@NonNull XmlContext context, @NonNull Element element) {
        if (element.getTagName().equals(AUTOMOTIVE_APP_TAG)) {
            NodeList children = element.getChildNodes();
            for (int i = 0; i < children.getLength(); i++) {
                Node childNode = children.item(i);
                if (childNode instanceof Element && ((Element) childNode).getTagName().equals(USES_TAG)) {
                    Element child = (Element) childNode;
                    String nameAttrValue = child.getAttribute(NAME_ATTRIBUTE);
                    if (!isValidNameAttribute(nameAttrValue)) {
                        context.report(ISSUE, child,
                                context.getLocation(child),
                                "The 'uses' element in <automotiveApp> should contain a valid value for the 'name' attribute. Valid values are 'media', 'notification', or 'sms'.",
                                null);
                    }
                }
            }
        }
    }

    private boolean isValidNameAttribute(@NonNull String nameValue) {
        return "media".equals(nameValue)
                || "notification".equals(nameValue)
                || "sms".equals(nameValue);
    }
}

File: lintgen/generated_agent/qwen-full/qwen2.5:32b-instruct/base+docs_agent/ResourceCycleDetector:ResourceCycle/iter_8.kt
package com.android.tools.lint.checks

import com.android.SdkConstants.*
import com.android.resources.ResourceFolderType
import com.android.tools.lint.detector.api.*
import org.w3c.dom.Element
import java.util.*

/**
 * Detects cycles in resource definitions which can lead to runtime exceptions.
 */
class ResourceCycleDetector : ResourceXmlDetector(), SourceCodeScanner {

    companion object {
        val ISSUE = Issue.create(
            id = "ResourceCycle",
            briefDescription = "Cycle in resource definitions",
            explanation = """
                Cycles in resource definitions can cause runtime exceptions. Ensure that there are no circular references between resources.
            """.trimIndent(),
            category = Category.CORRECTNESS,
            priority = 6,
            severity = Severity.ERROR,
            implementation = Implementation(
                ResourceCycleDetector::class.java,
                Scope.RESOURCE_FILE_SCOPE
            )
        )

        private val TAGS_TO_CHECK = listOf(TAG_STRING, TAG_PLURALS)
    }

    override fun appliesTo(folderType: ResourceFolderType): Boolean {
        return folderType == ResourceFolderType.VALUES
    }

    override fun visitResource(resFile: XmlContext) {
        if (resFile.file.name != null && resFile.rootElement.tagName in TAGS_TO_CHECK) {
            val graph = mutableMapOf<String, MutableList<String>>()
            buildGraph(resFile, graph)
            checkForCycles(graph, resFile)
        }
    }

    private fun buildGraph(file: XmlContext, graph: MutableMap<String, MutableList<String>>) {
        file.lintClient.getFiles(ResourceFolderType.VALUES).forEach { resFile ->
            val content = resFile.content
            if (content.documentElement != null) {
                traverseElement(content.documentElement as Element, graph)
            }
        }
    }

    private fun traverseElement(element: Element?, graph: MutableMap<String, MutableList<String>>) {
        element ?: return

        val name = element.getAttribute(ATTR_NAME)
        if (name.isNotEmpty()) {
            val references = mutableListOf<String>()
            for (i in 0 until element.attributes.length) {
                val attrName = element.attributes.item(i).nodeName
                if (attrName == ATTR_NAME || attrName == ATTR_ID) continue

                val refValue = element.getAttribute(attrName)
                if (refValue.startsWith("@")) {
                    references.add(refValue.substring(1))
                }
            }

            graph[name] = references
        }

        for (i in 0 until element.childNodes.length) {
            traverseElement(element.childNodes.item(i) as? Element, graph)
        }
    }

    private fun checkForCycles(graph: MutableMap<String, MutableList<String>>, resFile: XmlContext) {
        val visited = mutableSetOf<String>()
        val recursionStack = mutableSetOf<String>()

        for (node in graph.keys) {
            if (!visited.contains(node)) {
                if (isCyclicUtil(node, visited, recursionStack, graph)) {
                    reportIssue(graph, node, resFile)
                }
            }
        }
    }

    private fun isCyclicUtil(
        node: String,
        visited: MutableSet<String>,
        recursionStack: MutableSet<String>,
        graph: Map<String, MutableList<String>>
    ): Boolean {
        if (recursionStack.contains(node)) return true
        if (visited.contains(node)) return false

        visited.add(node)
        recursionStack.add(node)

        for (neighbor in graph[node] ?: emptyList()) {
            if (isCyclicUtil(neighbor, visited, recursionStack, graph)) {
                return true
            }
        }

        recursionStack.remove(node)
        return false
    }

    private fun reportIssue(graph: Map<String, MutableList<String>>, node: String, resFile: XmlContext) {
        val issue = ISSUE
        for ((key, value) in graph) {
            if (value.contains(node)) {
                val file = resFile.file
                if (file != null) {
                    val location = Location.create(file).message("Cycle detected involving $node")
                    report(resFile.context, issue, location)
                }
            }
        }
    }

    override fun getApplicableElements(): Set<String> =
        TAGS_TO_CHECK.toSet()
}